src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -455,94 +457,175 @@ static uct_iface_internal_ops_t uct_cuda_copy_iface_internal_ops = {
-        return status;
+        cuCtxPushCurrent(ctx_rsc->cuda_ctx);
+        ucs_memory_type_for_each(src) {
+            ucs_memory_type_for_each(dst) {
+                stream  = &ctx_rsc->queue_desc[src][dst].stream;
+                event_q = &ctx_rsc->queue_desc[src][dst].event_queue;
+
+                if (!ucs_queue_is_empty(event_q)) {
+                    ucs_warn("stream destroyed but queue not empty");
+                }
+
+                if (*stream == 0) {
+                    continue;
+                }
+
+                UCT_CUDADRV_FUNC_LOG_ERR(cuStreamDestroy(*stream));
+            }
+        }
+
+        if (ctx_rsc->short_stream) {
+            UCT_CUDADRV_FUNC_LOG_ERR(cuStreamDestroy(ctx_rsc->short_stream));
+        }
+
+        ucs_mpool_cleanup(&ctx_rsc->cuda_event_desc, 1);
+        cuCtxPopCurrent(&popped_ctx);
     }
 
-    self->id                     = ucs_generate_uuid((uintptr_t)self);
-    self->config.max_poll        = config->max_poll;
-    self->config.max_cuda_events = config->max_cuda_events;
-    self->config.bandwidth       = config->bandwidth;
-    UCS_BITMAP_CLEAR(&self->streams_to_sync);
+}
+
+ucs_status_t uct_cuda_copy_init_per_ctx_rscs(uct_cuda_copy_iface_t *iface,
+                                             CUcontext cuda_ctx,
+                                             unsigned long long ctx_id,
+                                             uct_cuda_copy_per_ctx_rsc_t *ctx_rsc)
+{
+    ucs_status_t status;
+    ucs_mpool_params_t mp_params;
+    ucs_memory_type_t src, dst;
 
     ucs_mpool_params_reset(&mp_params);
     mp_params.elem_size       = sizeof(uct_cuda_copy_event_desc_t);
     mp_params.elems_per_chunk = 128;
-    mp_params.max_elems       = self->config.max_cuda_events;
+    mp_params.priv_size       = sizeof(CUcontext);
+    mp_params.max_elems       = iface->config.max_cuda_events;
     mp_params.ops             = &uct_cuda_copy_event_desc_mpool_ops;
     mp_params.name            = "CUDA EVENT objects";
-    status = ucs_mpool_init(&mp_params, &self->cuda_event_desc);
+    status = ucs_mpool_init(&mp_params, &ctx_rsc->cuda_event_desc);
     if (UCS_OK != status) {
         ucs_error("mpool creation failed");
         return UCS_ERR_IO_ERROR;
     }
 
-    ucs_queue_head_init(&self->active_queue);
-
     ucs_memory_type_for_each(src) {
         ucs_memory_type_for_each(dst) {
-            self->queue_desc[src][dst].stream = 0;
-            ucs_queue_head_init(&self->queue_desc[src][dst].event_queue);
+            ctx_rsc->queue_desc[src][dst].stream = 0;
+            ucs_queue_head_init(&ctx_rsc->queue_desc[src][dst].event_queue);
         }
     }
 
-    self->short_stream = 0;
-    self->cuda_context = 0;
+    ctx_rsc->short_stream = 0;
+    ctx_rsc->cuda_ctx     = cuda_ctx;
+    ctx_rsc->ctx_id       = ctx_id;
 
     return UCS_OK;
 }
 
-static UCS_CLASS_CLEANUP_FUNC(uct_cuda_copy_iface_t)
+ucs_status_t uct_cuda_copy_get_ctx_rscs(uct_cuda_copy_iface_t *iface,
```

Comment:
Should this have `_per_` in the name to be consistent with `init_per_ctx_rscs` and `cleanup_per_ctx_rscs` functions?